Removed some features + Balance + CrossVal

In [1]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import math
import IPython


# visualization
import seaborn as sns
sns.set(style="white")  
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')
from pandas.plotting import scatter_matrix

# warnings handler
import warnings
warnings.filterwarnings("ignore")

# Machine Learning Libraries
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import fbeta_score, accuracy_score, f1_score, recall_score, precision_score
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
#from xgboost import XGBClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer 
from sklearn.model_selection import KFold
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.decomposition import PCA

# Pipeline
from sklearn.model_selection import train_test_split, cross_val_predict
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer

In [2]:
import statsmodels.api as sm

In [3]:
df = pd.read_csv("/Users/kevintomas/Desktop/nf-sep-20/capstone kram/Data/no_dupli_and_outl.csv")

In [4]:
df.drop("Unnamed: 0", axis=1, inplace=True)

In [5]:
print(df.shape)
df.drop(["avg_churn", 
 "ort", 
 "date_x", 
 "kuendigungs_eingangs_datum", 
 "abo_registrierung_min", 
 "training_set", 
 "liefer_beginn_evt", 
 "nl_registrierung_min",
 "zon_rawr",
 "zon_community",
 "zon_app_sonstige",
 "zon_schach",
 "zon_blog_kommentare",
 "zon_quiz",
 "zon_boa",
 "zon_kommentar",
 'plz_1',
 'plz_2',
 'cnt_abo_diezeit',
 'cnt_abo_diezeit_digital',
 'cnt_abo_magazin',
 'received_anzahl_1w',
 'received_anzahl_1m',
 'received_anzahl_3m',
 'opened_anzahl_1w',
 'opened_anzahl_1m',
 'opened_anzahl_3m',
 'clicked_anzahl_1w',
 'clicked_anzahl_1m',
 'clicked_anzahl_3m',
 'unsubscribed_anzahl_1w',
 'unsubscribed_anzahl_1m',
 'unsubscribed_anzahl_3m',
 'openrate_1w',
 'clickrate_1w',
 'openrate_1m',
 'clickrate_1m',
 'received_anzahl_bestandskunden_1w',
 'received_anzahl_bestandskunden_1m',
 'received_anzahl_bestandskunden_3m',
 'opened_anzahl_bestandskunden_1w',
 'opened_anzahl_bestandskunden_1m',
 'opened_anzahl_bestandskunden_3m',
 'clicked_anzahl_bestandskunden_1w',
 'clicked_anzahl_bestandskunden_1m',
 'clicked_anzahl_bestandskunden_3m',
 'unsubscribed_anzahl_bestandskunden_1w',
 'unsubscribed_anzahl_bestandskunden_1m',
 'unsubscribed_anzahl_bestandskunden_3m',
 'openrate_bestandskunden_1w',
 'clickrate_bestandskunden_1w',
 'openrate_bestandskunden_1m',
 'clickrate_bestandskunden_1m',
 'received_anzahl_produktnews_1w',
 'received_anzahl_produktnews_1m',
 'received_anzahl_produktnews_3m',
 'opened_anzahl_produktnews_1w',
 'opened_anzahl_produktnews_1m',
 'opened_anzahl_produktnews_3m',
 'clicked_anzahl_produktnews_1w',
 'clicked_anzahl_produktnews_1m',
 'clicked_anzahl_produktnews_3m',
 'unsubscribed_anzahl_produktnews_1w',
 'unsubscribed_anzahl_produktnews_1m',
 'unsubscribed_anzahl_produktnews_3m',
 'openrate_produktnews_1w',
 'clickrate_produktnews_1w',
 'openrate_produktnews_1m',
 'clickrate_produktnews_1m',
 'received_anzahl_hamburg_1w',
 'received_anzahl_hamburg_1m',
 'received_anzahl_hamburg_3m',
 'opened_anzahl_hamburg_1w',
 'opened_anzahl_hamburg_1m',
 'opened_anzahl_hamburg_3m',
 'clicked_anzahl_hamburg_1w',
 'clicked_anzahl_hamburg_1m',
 'clicked_anzahl_hamburg_3m',
 'unsubscribed_anzahl_hamburg_1w',
 'unsubscribed_anzahl_hamburg_1m',
 'unsubscribed_anzahl_hamburg_3m',
 'openrate_hamburg_1w',
 'clickrate_hamburg_1w',
 'openrate_hamburg_1m',
 'clickrate_hamburg_1m',
 'received_anzahl_zeitbrief_1w',
 'received_anzahl_zeitbrief_1m',
 'received_anzahl_zeitbrief_3m',
 'opened_anzahl_zeitbrief_1w',
 'opened_anzahl_zeitbrief_1m',
 'opened_anzahl_zeitbrief_3m',
 'clicked_anzahl_zeitbrief_1w',
 'clicked_anzahl_zeitbrief_1m',
 'clicked_anzahl_zeitbrief_3m',
 'unsubscribed_anzahl_zeitbrief_1w',
 'unsubscribed_anzahl_zeitbrief_1m',
 'unsubscribed_anzahl_zeitbrief_3m',
 'openrate_zeitbrief_1w',
 'clickrate_zeitbrief_1w',
 'openrate_zeitbrief_1m',
 'clickrate_zeitbrief_1m',], axis=1, inplace=True)
print(df.shape)

(143053, 169)
(143053, 68)


In [6]:
df=df.dropna()

**Balancing**

In [7]:
count_class_0, count_class_1 = df.churn.value_counts()

In [8]:
churn_class_0 = df[df['churn'] == 0]
churn_class_1 = df[df['churn'] == 1]

**Upsampling**

In [9]:
churn_class_1_up = churn_class_1.sample(count_class_0, replace=True)

In [10]:
df = pd.concat([churn_class_1_up, churn_class_0], axis=0)

In [180]:
df.churn.value_counts()

1    99701
0    99701
Name: churn, dtype: int64

**Downsampling**

In [96]:
churn_class_1_down = churn_class_0.sample(count_class_1, replace=True)

In [97]:
df = pd.concat([churn_class_1_down, churn_class_1], axis=0)

In [98]:
df.churn.value_counts()

1    43340
0    43340
Name: churn, dtype: int64

In [11]:
y = df['churn']
X = df.drop('churn', axis = 1)
X.shape

(199402, 67)

**Dummies**

In [12]:
other = ["zon_che_opt_in", 
 "zon_sit_opt_in",
 "zon_zp_grey",
 "zon_premium",
 "zon_sonstige",
 "zon_zp_red",
 "nl_zeitbrief",
 "nl_zeitshop",
 "nl_zeitverlag_hamburg",
 "nl_fdz_organisch"]

In [13]:
categoric_features = list(X.columns[X.dtypes==object]) + other

In [14]:
categoric_features

['kanal',
 'objekt_name',
 'aboform_name',
 'zahlung_rhythmus_name',
 'zahlung_weg_name',
 'plz_3',
 'land_iso_code',
 'anrede',
 'titel',
 'zon_che_opt_in',
 'zon_sit_opt_in',
 'zon_zp_grey',
 'zon_premium',
 'zon_sonstige',
 'zon_zp_red',
 'nl_zeitbrief',
 'nl_zeitshop',
 'nl_zeitverlag_hamburg',
 'nl_fdz_organisch']

In [15]:
dummies = pd.get_dummies(X, columns=categoric_features, drop_first=True)

**Scaling**

In [16]:
numeric_features = ['lesedauer',
 'shop_kauf',
 'cnt_abo',
 'cnt_umwandlungsstatus2_dkey',
 'nl_blacklist_dum',
 'nl_bounced_dum',
 'nl_aktivitaet',
 'nl_sperrliste_dum',
 'nl_opt_in_dum',
 'received_anzahl_6m',
 'openedanzahl_6m',
 'clicked_anzahl_6m',
 'unsubscribed_anzahl_6m',
 'openrate_3m',
 'clickrate_3m',
 'received_anzahl_bestandskunden_6m',
 'openedanzahl_bestandskunden_6m',
 'clicked_anzahl_bestandskunden_6m',
 'unsubscribed_anzahl_bestandskunden_6m',
 'openrate_bestandskunden_3m',
 'clickrate_bestandskunden_3m',
 'received_anzahl_produktnews_6m',
 'openedanzahl_produktnews_6m',
 'clicked_anzahl_produktnews_6m',
 'unsubscribed_anzahl_produktnews_6m',
 'openrate_produktnews_3m',
 'clickrate_produktnews_3m',
 'received_anzahl_hamburg_6m',
 'openedanzahl_hamburg_6m',
 'clicked_anzahl_hamburg_6m',
 'unsubscribed_anzahl_hamburg_6m',
 'openrate_hamburg_3m',
 'clickrate_hamburg_3m',
 'received_anzahl_zeitbrief_6m',
 'openedanzahl_zeitbrief_6m',
 'clicked_anzahl_zeitbrief_6m',
 'unsubscribed_anzahl_zeitbrief_6m',
 'openrate_zeitbrief_3m',
 'clickrate_zeitbrief_3m']

In [17]:
scaler = StandardScaler()

In [18]:
X_scaled = scaler.fit_transform(dummies[numeric_features])

In [19]:
X_preprocessed = np.concatenate([X_scaled, dummies.drop(numeric_features, axis=1)], axis=1)

**Splitting**

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size = .3, random_state=25)

In [21]:
X_train.shape

(139581, 794)

## PCA

In [87]:
pca = PCA(n_components=None)

In [88]:
X_train_scaled_pca  = pca.fit(X_train)

In [89]:
X_train_trans_pca = pca.transform(X_train)

In [90]:
pca_explained_var = pca.explained_variance_ratio_

cum_explaiend_var = pca_explained_var.cumsum()

In [91]:
sum_variance, component_count = 0, 0
while sum_variance < 0.50:
    sum_variance += pca.explained_variance_ratio_[component_count]
    component_count += 1
    
print('Number of Principal Components that explain >=50% of Variance: ', component_count)
print('Total Variance Explained by '+str(component_count)+' components:', str(sum_variance*100)+'%')

Number of Principal Components that explain >=50% of Variance:  9
Total Variance Explained by 9 components: 51.65510125713676%


## Modelling with 9 components explaining 50% of the variance

In [101]:
pca = PCA(n_components=9)
pca.fit(X_train)

PCA(n_components=9)

In [102]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [103]:
X_train_pca_cst = sm.add_constant(X_train_pca)
X_test_pca_cst = sm.add_constant(X_test_pca)

**LOGREG**

In [95]:
model = LogisticRegression().fit(X_train_pca_cst, y_train)

In [96]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [97]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 63.812409998495504%
Testing Set Accuracy: 63.97251801206934%
Training Set Recall: 69.51194990233253%
Testing Set Recall: 69.84406689496959%
Training Set Precision: 62.30223606802178%
Testing Set Precision: 62.72805995998686%


**RANDOM FOREST**

In [98]:
model = RandomForestClassifier().fit(X_train_pca_cst, y_train)

In [99]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [100]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.987820691928%
Testing Set Accuracy: 85.40479095969643%
Training Set Recall: 99.99425485464782%
Testing Set Recall: 88.43634671011071%
Training Set Precision: 99.98133069090804%
Testing Set Precision: 83.50809996232576%


<br/>**CROSSVALIDATION WITH 50% VARIANCE**<br/>

In [42]:
from sklearn.model_selection import cross_val_score

**RANDOMFOREST**

**Upsampled**

In [104]:
clf = RandomForestClassifier()
recall = cross_val_score(clf, X_train_pca_cst, y_train, cv=5, scoring="recall")

In [105]:
print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Recall: 0.85 (+/- 0.01)


## Modelling with 20 components explaining 75% of the variance

In [78]:
pca = PCA(n_components=20)
pca.fit(X_train)

PCA(n_components=20)

In [79]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [80]:
X_train_pca_cst = sm.add_constant(X_train_pca)
X_test_pca_cst = sm.add_constant(X_test_pca)

**LOGREG**

In [81]:
model = LogisticRegression().fit(X_train_pca_cst, y_train)

In [82]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [83]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 64.88992054792557%
Testing Set Accuracy: 64.97216696477825%
Training Set Recall: 69.31230610134436%
Testing Set Recall: 69.61798051667387%
Training Set Precision: 63.58186537372033%
Testing Set Precision: 63.92611814990078%


**RANDOM FOREST**

In [84]:
model = RandomForestClassifier().fit(X_train_pca_cst, y_train)

In [85]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [86]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.9871042620414%
Testing Set Accuracy: 86.22222965179452%
Training Set Recall: 99.99281856830979%
Testing Set Recall: 88.85859626957476%
Training Set Precision: 99.98133042279412%
Testing Set Precision: 84.53047411202834%


## Modelling with 28 components explaining 85% of the variance

In [57]:
pca = PCA(n_components=28)
pca.fit(X_train)

PCA(n_components=28)

In [58]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [59]:
X_train_pca_cst = sm.add_constant(X_train_pca)
X_test_pca_cst = sm.add_constant(X_test_pca)

**LOGREG**

In [30]:
model = LogisticRegression().fit(X_train_pca_cst, y_train)

In [31]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [32]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 65.61996260235992%
Testing Set Accuracy: 65.40846859798398%
Training Set Recall: 67.8142594507641%
Testing Set Recall: 67.81261428998903%
Training Set Precision: 64.86111492705443%
Testing Set Precision: 64.93886907794193%


**Upsampled**<br/>
Training Set Accuracy: 65.6744112737407%<br/>
Testing Set Accuracy: 65.4251851356547%<br/>
Training Set Recall: 67.84585775020108%<br/>
Testing Set Recall: 67.68959670179872%<br/>
Training Set Precision: 64.91987575932492%<br/>
Testing Set Precision: 64.99489209551781%<br/>

**Downsampled**<br/>
Training Set Accuracy: 65.51519546443404%<br/>
Testing Set Accuracy: 65.39378557145055%<br/>
Training Set Recall: 67.2543247444845%<br/>
Testing Set Recall: 66.93369954985884%<br/>
Training Set Precision: 64.84357559715535%<br/>
Testing Set Precision: 65.28501265069207%<br/>

**RANDOM FOREST**

In [60]:
model = RandomForestClassifier().fit(X_train_pca_cst, y_train)

In [61]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [62]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.9871042620414%
Testing Set Accuracy: 86.34760368432491%
Training Set Recall: 99.99425485464782%
Testing Set Recall: 88.68903148585298%
Training Set Precision: 99.97989487893845%
Testing Set Precision: 84.8441475826972%


## Modelling with 28 components explaining 85% of the variance

In [57]:
pca = PCA(n_components=28)
pca.fit(X_train)

PCA(n_components=28)

In [58]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [59]:
X_train_pca_cst = sm.add_constant(X_train_pca)
X_test_pca_cst = sm.add_constant(X_test_pca)

**LOGREG**

In [30]:
model = LogisticRegression().fit(X_train_pca_cst, y_train)

In [31]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [32]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 65.61996260235992%
Testing Set Accuracy: 65.40846859798398%
Training Set Recall: 67.8142594507641%
Testing Set Recall: 67.81261428998903%
Training Set Precision: 64.86111492705443%
Testing Set Precision: 64.93886907794193%


**Upsampled**<br/>
Training Set Accuracy: 65.6744112737407%<br/>
Testing Set Accuracy: 65.4251851356547%<br/>
Training Set Recall: 67.84585775020108%<br/>
Testing Set Recall: 67.68959670179872%<br/>
Training Set Precision: 64.91987575932492%<br/>
Testing Set Precision: 64.99489209551781%<br/>

**Downsampled**<br/>
Training Set Accuracy: 65.51519546443404%<br/>
Testing Set Accuracy: 65.39378557145055%<br/>
Training Set Recall: 67.2543247444845%<br/>
Testing Set Recall: 66.93369954985884%<br/>
Training Set Precision: 64.84357559715535%<br/>
Testing Set Precision: 65.28501265069207%<br/>

**RANDOM FOREST**

In [60]:
model = RandomForestClassifier().fit(X_train_pca_cst, y_train)

In [61]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [62]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.9871042620414%
Testing Set Accuracy: 86.34760368432491%
Training Set Recall: 99.99425485464782%
Testing Set Recall: 88.68903148585298%
Training Set Precision: 99.97989487893845%
Testing Set Precision: 84.8441475826972%


## Modelling with 50 components explaining 95% of the variance

In [33]:
pca = PCA(n_components=50)
pca.fit(X_train)

PCA(n_components=50)

In [34]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [35]:
X_train_pca_cst = sm.add_constant(X_train_pca)
X_test_pca_cst = sm.add_constant(X_test_pca)

Training

In [36]:
model = LogisticRegression().fit(X_train_pca_cst, y_train)

In [37]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [38]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 67.76423725292125%
Testing Set Accuracy: 67.50639407565905%
Training Set Recall: 69.17011375387797%
Testing Set Recall: 68.92642218306348%
Training Set Precision: 67.1776702143983%
Testing Set Precision: 67.2582162670733%


**Upsampled**<br/>
Training Set Accuracy: 67.70692286199412%<br/>
Testing Set Accuracy: 67.50472242189198%<br/>
Training Set Recall: 68.92307250373435%<br/>
Testing Set Recall: 68.85660138976627%<br/>
Training Set Precision: 67.1851592579629%<br/>
Testing Set Precision: 67.27957897472548%<br/>

**Downsampled**<br/>
Training Set Accuracy: 67.58191047531149%<br/>
Testing Set Accuracy: 67.59729272419628%<br/>
Training Set Recall: 68.72953395296531%<br/>
Testing Set Recall: 68.2383459220264%<br/>
Training Set Precision: 67.03984513631231%<br/>
Testing Set Precision: 67.72166275459983%<br/>

**RANDOM FOREST**

In [39]:
model = RandomForestClassifier().fit(X_train_pca_cst, y_train)

In [40]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [41]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.9871042620414%
Testing Set Accuracy: 87.63644873873724%
Training Set Recall: 99.99138228197172%
Testing Set Recall: 88.60591149383251%
Training Set Precision: 99.98276604911676%
Testing Set Precision: 87.03746040040498%


**Upsampled**<br/>
Training Set Accuracy: 99.9871042620414%<br/>
Testing Set Accuracy: 87.26366994868023%<br/>
Training Set Recall: 99.99856371366195%<br/>
Testing Set Recall: 87.7414635768195%<br/>
Training Set Precision: 99.97558874210225%<br/>
Testing Set Precision: 87.03251764395489%<br/>

**Downsampled**<br/>
Training Set Accuracy: 99.98846331333641%<br/>
Testing Set Accuracy: 71.7005076142132%<br/>
Training Set Recall: 99.99338471206958%<br/>
Testing Set Recall: 68.29938200961318%<br/>
Training Set Precision: 99.98346342108745%<br/>
Testing Set Precision: 73.64264560710761%<br/>

**RANDOM FOREST MIT PARAMETERN VON ANDERER GRUPPE**

In [51]:
model = RandomForestClassifier(n_estimators = 380, criterion = "entropy", max_depth = 35,
                                     max_features = "auto", min_samples_leaf = 1,
                                     min_samples_split = 4, n_jobs = -1, bootstrap = True).fit(X_train_pca_cst, y_train)

In [52]:
y_pred_train = model.predict(X_train_pca_cst)
y_pred_test = model.predict(X_test_pca_cst)

In [53]:
train_score = accuracy_score(y_train, y_pred_train)*100
test_score = accuracy_score(y_test, y_pred_test)*100
recall_train = recall_score(y_train, y_pred_train)*100
recall_test = recall_score(y_test, y_pred_test)*100
precision_train = precision_score(y_train, y_pred_train)*100
precision_test = precision_score(y_test, y_pred_test)*100


print("Training Set Accuracy:",str(train_score)+'%')
print("Testing Set Accuracy:",str(test_score)+'%')
print("Training Set Recall:",str(recall_train)+'%')
print("Testing Set Recall:",str(recall_test)+'%')
print("Training Set Precision:",str(precision_train)+'%')
print("Testing Set Precision:",str(precision_test)+'%')

Training Set Accuracy: 99.9871042620414%
Testing Set Accuracy: 87.5077313986727%
Training Set Recall: 99.98994599563369%
Testing Set Recall: 89.05475945074309%
Training Set Precision: 99.98420175791348%
Testing Set Precision: 86.49809468449267%


## Crossvalidation

In [42]:
from sklearn.model_selection import cross_val_score

LOGREG

In [43]:
clf = LogisticRegression()
scores = cross_val_score(clf, X_train_pca_cst, y_train, cv=5, scoring="recall")

In [44]:
print("Recall: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Recall: 0.69 (+/- 0.00)


**RANDOMFOREST**

**Downsampled**

In [45]:
clf = RandomForestClassifier()
recall = cross_val_score(clf, X_train_pca_cst, y_train, cv=5, scoring="recall")

In [46]:
print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Recall: 0.85 (+/- 0.01)


**Upsampled**

In [163]:
clf = RandomForestClassifier()
scores = cross_val_score(clf, X_train_pca_cst, y_train, cv=5)

In [164]:
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.84 (+/- 0.01)


In [55]:
clf = RandomForestClassifier(n_estimators = 380, criterion = "entropy", max_depth = 35,
                                     max_features = "auto", min_samples_leaf = 1,
                                     min_samples_split = 4, n_jobs = -1, bootstrap = True)
                             
recall = cross_val_score(clf, X_train_pca_cst, y_train, cv=5, scoring="recall")

In [56]:
print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Recall: 0.85 (+/- 0.01)


In [170]:
clf = RandomForestClassifier()
recall = cross_val_score(clf, X_train_pca_cst, y_train, cv=5, scoring="precision")

In [171]:
print("Precision: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Precision: 0.83 (+/- 0.00)


**include all data**

In [190]:
pca = PCA(n_components=None)

In [192]:
X_scaled_pca  = pca.fit(X_preprocessed)

In [193]:
X_trans_pca = pca.transform(X_preprocessed)

In [194]:
pca_explained_var = pca.explained_variance_ratio_

cum_explaiend_var = pca_explained_var.cumsum()

In [195]:
sum_variance, component_count = 0, 0
while sum_variance < 0.95:
    sum_variance += pca.explained_variance_ratio_[component_count]
    component_count += 1
    
print('Number of Principal Components that explain >=95% of Variance: ', component_count)
print('Total Variance Explained by '+str(component_count)+' components:', str(sum_variance*100)+'%')

Number of Principal Components that explain >=95% of Variance:  50
Total Variance Explained by 50 components: 95.03986697695774%


In [196]:
pca = PCA(n_components=50)
pca.fit(X_train)

PCA(n_components=50)

In [197]:
X_pca = pca.transform(X_preprocessed)

In [199]:
X_pca_cst = sm.add_constant(X_pca)

In [200]:
clf = RandomForestClassifier()
scores = cross_val_score(clf, X_pca_cst, y, cv=7)

In [201]:
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.87 (+/- 0.14)


In [202]:
clf = RandomForestClassifier()
recall = cross_val_score(clf, X_pca_cst, y, cv=7, scoring="recall")

In [203]:
print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Recall: 0.91 (+/- 0.01)


In [ ]:
clf = RandomForestClassifier()
precision = cross_val_score(clf, X_pca_cst, y, cv=7, scoring="precision")

In [ ]:
print("Precision: %0.2f (+/- %0.2f)" % (precision.mean(), precision.std() * 2))